src/libsyntax/parse/parser.rs

```diff
@@ -3812,22 +3802,18 @@ impl<'a> Parser<'a> {
         Ok(self.mk_expr(lo.to(hi), ExprKind::If(cond, thn, els), attrs))
     }
 
-    /// Parses an `if let` expression (`if` token already eaten).
-    fn parse_if_let_expr(&mut self, attrs: ThinVec<Attribute>)
-                             -> PResult<'a, P<Expr>> {
+    /// Parses a `let $pats = $expr` pseudo-expression.
+    /// The `let` token has already been eaten.
+    fn parse_let_expr(&mut self, attrs: ThinVec<Attribute>) -> PResult<'a, P<Expr>> {
         let lo = self.prev_span;
-        self.expect_keyword(keywords::Let)?;
         let pats = self.parse_pats()?;
         self.expect(&token::Eq)?;
-        let expr = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
-        let thn = self.parse_block()?;
-        let (hi, els) = if self.eat_keyword(keywords::Else) {
-            let expr = self.parse_else_expr()?;
-            (expr.span, Some(expr))
-        } else {
-            (thn.span, None)
-        };
-        Ok(self.mk_expr(lo.to(hi), ExprKind::IfLet(pats, expr, thn, els), attrs))
+
+        let expr = self.with_res(
+            Restrictions::NO_STRUCT_LITERAL,
+            |this| this.parse_assoc_expr_with(1 + AssocOp::LAnd.precedence(), None.into())
```

Comment:
N.B. The logic here is that we force parenthesis for anything with a precedence <= that of `&&`. This [diverges slightly from the RFC](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#reference-level-explanation) but I think that is for the better as the RFC specifies thing in a rather hackish way that would require statefully tracking whether we are inside the condition of an `if` and changing the precedence table based on that. Instead, this PR uses a much simpler approach.

Reply:
Isn't it actually `<` than `&&`? Or what's with the `1 +`?

Reply:
No it is `<= &&` and that's what the `1 +` achieves because we want everything with `> &&` to not require parenthesis. In other words, we want to force `if let a = b && c` to associate as `if (let a = b) && c`. If you remove the `1 +` then that won't happen.

Reply:
To understand the logic of this, you'll want to take a look at the definition of `fn parse_assoc_expr_with`. Specifically:
```rust
            let prec = op.precedence();
            if prec < min_prec {
                break;
            }
```

By giving `min_prec = 1 + AssocOp::LAnd.precedence()` and `op = AssocOp::LAnd` we will have:
```rust
if AssocOp::LAnd.precedence() < 1 + AssocOp::LAnd.precedence() {
    break;
}

==>

if true {
    break;
}

==>

break;
```